cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/ResourceDistributionGoal.java

```diff
@@ -333,11 +340,11 @@ protected void rebalanceForBroker(Broker broker,
     boolean unbalanced = false;
     if (requireLessLoad) {
       if (rebalanceByMovingLoadOut(broker, clusterModel, optimizedGoals, REPLICA_MOVEMENT, optimizationOptions)) {
-        unbalanced = rebalanceBySwappingLoadOut(broker, clusterModel, optimizedGoals, optimizationOptions);
+        unbalanced = rebalanceBySwappingLoadOut(broker, clusterModel, optimizedGoals, optimizationOptions, moveImmigrantsOnly);
       }
     } else if (requireMoreLoad) {
-      if (rebalanceByMovingLoadIn(broker, clusterModel, optimizedGoals, REPLICA_MOVEMENT, optimizationOptions)) {
-        unbalanced = rebalanceBySwappingLoadIn(broker, clusterModel, optimizedGoals, optimizationOptions);
+      if (rebalanceByMovingLoadIn(broker, clusterModel, optimizedGoals, REPLICA_MOVEMENT, optimizationOptions, moveImmigrantsOnly)) {
```

Comment:
Offline replica on offline logDirs? If that is the case, it is online relevant to `2.0` branch. Also we cannot swap with an offline replica, because the destination disk or broker would be down. Finally, the `this#_selfHealingDeadBrokersOnly` ensures that there will be no offline replicas on dead brokers. Please let me know if I am missing something.

Reply:
Actually I originally thought the scenario that in a cluster with dead broker, and we first optimize for a healthy broker and it needs more load, can it get some offline replica from the dead broker here? Now I see it only try to get replica from alive broker, then it make sense.
And you are right about swap operation.